### PR TITLE
Remove memory allocation and replace with slices

### PR DESCRIPTION
**Pull request by SCingolani**
This change completely removes the memory management from Heimlig, instead relying on slices provided in the requests for both input and output parameters.

As part of this change, the `Channel` concept was replaced with a `Source` + `Sink` abstraction which lends itself better to a future use of the async `Stream` and `Sink` traits and simplifies the handling of lifetimes.

This change stems from a discussion in #156 where an alternative abstraction over parameters to and from Heimlig was suggested. Compared to that, this change leverages the full power of the borrow checker to make sure that neither the internal Heimlig code, or the integration code mishandles the memory used to pass in parameters and write out results.

### CODE REVIEW

File: heimlig/src/client/api.rs [hunk 1]
```diff
@@ -6,95 +6,113 @@ pub enum Error {
...
+            requests_sink,
+            responses_source,
+        }
     }
 
     /// Request `size` many random bytes.
-    pub fn get_random(&mut self, size: usize) -> Result<(), Error> {
-        self.channel.send(Request::GetRandom { size })
+    pub fn get_random(&mut self, output: &'a mut [u8]) -> Result<(), Error> {
+        self.requests_sink.send(Request::GetRandom { output })
```
Comment (SCingolani) [inline heimlig/src/client/api.rs:33]:
```suggestion
        if self.requests_sink.ready() {
            self.requests_sink.send(Request::GetRandom { output })
        } else {
            Err(Error::QueueFull)
        }
```

File: heimlig/src/hsm/workers/chachapoly_worker.rs [hunk 2]
```diff
@@ -1,83 +1,70 @@
 use crate::common::jobs::{Error, Response};
-use crate::common::pool::{Pool, PoolChunk};
 
-pub struct ChachaPolyWorker<'a> {
-    pub pool: &'a Pool,
-}
+pub struct ChachaPolyWorker {}
```
[the remark this answers is not in the document]
Comment (SCingolani) [reply heimlig/src/hsm/workers/chachapoly_worker.rs:3]: Arguably do we even need a type for this...

File: heimlig/src/hsm/workers/chachapoly_worker.rs [hunk 3]
```diff
@@ -1,83 +1,70 @@
...
+        aad: Option<&[u8]>,
+        plaintext: &'a mut [u8],
+        tag: &[u8],
+    ) -> Response<'a> {
         let aad = match &aad {
-            Some(aad) => aad.as_slice(),
+            Some(aad) => aad,
             None => &[] as &[u8],
         };
         match crate::crypto::chacha20poly1305::decrypt_in_place_detached(
```
Comment (SCingolani) [inline heimlig/src/hsm/workers/chachapoly_worker.rs:66]: shall we check tag size before calling decrypt_in_place_detached? or does it do a check internally? if so, what is the correct size?

Comment (rpiasetskyi) [reply heimlig/src/hsm/workers/chachapoly_worker.rs:66]: `decrypt_in_place_detached` checks the tag size.
https://github.com/eclipse-heimlig/heimlig/blob/cd3b0689f5e5a339b6aa1e3cd6ec37898a90d10a/heimlig/src/crypto/chacha20poly1305.rs#L60

The `TAG_SIZE` is 16 bytes.
https://github.com/eclipse-heimlig/heimlig/blob/cd3b0689f5e5a339b6aa1e3cd6ec37898a90d10a/heimlig/src/crypto/chacha20poly1305.rs#L12

File: heimlig/src/hsm/workers/chachapoly_worker.rs [hunk 4]
```diff
@@ -1,83 +1,70 @@
...
-                        tag.as_slice_mut().copy_from_slice(computed_tag.as_slice());
-                        Response::EncryptChaChaPoly { ciphertext, tag }
-                    }
-                    Err(e) => Response::Error(Error::Crypto(e)),
+        nonce: &[u8],
+        aad: Option<&[u8]>,
+        ciphertext: &'a mut [u8],
+        tag: &'a mut [u8],
+    ) -> Response<'a> {
+        let aad = match &aad {
```
[the remark this answers is not in the document]
Comment (SCingolani) [reply heimlig/src/hsm/workers/chachapoly_worker.rs]: Yes, done :)

File: heimlig/src/hsm/workers/chachapoly_worker.rs [hunk 5]
```diff
@@ -1,83 +1,70 @@
...
+            Some(aad) => aad,
+            None => &[] as &[u8],
+        };
+        match crate::crypto::chacha20poly1305::encrypt_in_place_detached(
+            key, nonce, aad, ciphertext,
+        ) {
+            Ok(computed_tag) => {
+                if computed_tag.len() != tag.len() {
+                    return Response::Error(Error::Alloc);
```
[the remark this answers is not in the document]
Comment (SCingolani) [reply heimlig/src/hsm/workers/chachapoly_worker.rs]: Fixed

File: heimlig/src/hsm/workers/chachapoly_worker.rs (hunk 3 above)
Comment (SCingolani) [reply heimlig/src/hsm/workers/chachapoly_worker.rs:61]: See proposed change in [f970059](https://github.com/eclipse-heimlig/heimlig/pull/157/commits/f9700595fa9ef8af1c5f43bfe401074a38b75170), i.e., make it explicit in the types that the tag should have the correct size; this means offloading the error handling to the user; see for instance [the test in scheduler](https://github.com/eclipse-heimlig/heimlig/pull/157/commits/f9700595fa9ef8af1c5f43bfe401074a38b75170#diff-bb6a5fb883712e7e170d36fd6dd4d53701e97fdd7a20acd6deea484098c8f452R108).

**Comment by SCingolani** [comment]
If we do [e62f76c](https://github.com/eclipse-heimlig/heimlig/pull/157/commits/e62f76c507990f611587e8aed34e62c1bee36475) then we should also remove the tag size check in `ChachaPolyWorker::encrypt`, and we could also propagate this change to `crypto/chacha20poly1305.rs`. If so, then I would say we move this change to another PR and change the tag to be of fixed sized everywhere in one single commit.

File: examples/linux/src/main.rs [hunk 6]
```diff
@@ -92,24 +103,22 @@ async fn hsm_task(
...
 
     loop {
         // Send request
         Timer::after(Duration::from_millis(1000)).await;
-        let random_size = 16;
-        info!(target: "CLIENT", "Sending request: random data (size={})", random_size);
-        api.get_random(random_size)
+        let random_output = Box::leak(Box::new([0u8; 16]));
+        info!(target: "CLIENT", "Sending request: random data (size={})", random_output.len());
+        api.get_random(random_output.as_mut_slice())
```
Comment (inorick) [inline examples/linux/src/main.rs:121]: Try prepending a `&mut` instead.

File: examples/linux/src/main.rs [hunk 7]
```diff
@@ -122,8 +131,10 @@ async fn client_task(
                             info!(target: "CLIENT",
                                 "Received response: random data (size={}): {}",
                                 data.len(),
-                                hex::encode(data.as_slice())
+                                hex::encode(&data)
                             );
+                            // release the memory
+                            drop(unsafe { Box::from_raw(data) });
```
Comment (inorick) [inline examples/linux/src/main.rs:137]: Verify that the memory returned matches the one handed out for the request.

File: examples/stm32h745i/cm7/src/bin/rng_single_core.rs [hunk 8]
```diff
@@ -108,28 +119,49 @@ async fn hsm_task(
...
-        let random_size = 16;
-        info!("Sending request: random data (size={})", random_size);
-        api.get_random(random_size)
+
+        let mut random_buffer_alloc = pool
+            .alloc()
+            .expect("Failed to allocate buffer for random data")
+            .init([0; 16]);
+        // Safety: we forget about the box below, so it doesn't get dropped!
+        let random_buffer = unsafe {
```
Comment (inorick) [inline examples/stm32h745i/cm7/src/bin/rng_single_core.rs:150]: Upstream as `leak` to `heapless` upstream.

**Pull request merged**